### ISSUE TITLE
Enable Heartbeat Post Locking

### ISSUE THREAD

**Issue by adamsilverstein** [WordPress/gutenberg#4331]
## Issue Overview
Enable heartbeat based post locking functionality.

Work with the existing editor functionality for post locking. Enhance when collaborative editing is available by offering 'Collaborate' as an additional option in the takeover modal and possibly post list options.

## Expected Behavior
* Post locking should work similar to or better than the classic editor's post locking feature: https://codex.wordpress.org/Post_Locking as well as respecting any locking of posts in the classic editor
* Posts that are being edited are locked to prevent multiple editors overwriting each others changes
* When you try to open the post the editor is disabled and a modal offers 'go back', 'preview' and 'take over', plus 'collaborate' if collaborative editing is available - the editor remains inactive/inaccessible unless you choose to take over or collaborate.
* When you are editing a post and another user takes over, a similar modal pops up. Currently you can no longer edit, the only option is to return to the posts list screen. Perhaps this should instead offer collaborate or take over?

## Testing
* Create and edit a post.
* In a separate browser, log in with a different user. Check post list screen and look for the post lock indicators.
* Test in the classic editor and Gutenberg.

## Current Behavior
* Posts do not auto lock.
* Opening a locked post from the classic editor in Gutenberg does not respect the post lock.
* Opening a locked post in the classic editor with the Gutenberg plugin active and choosing to take over the post does not work correctly - instead of the classic editor, I am taken to the Gutenberg editor.

## Screenshots

Some screenshots from the classic editor:

post list lock indicator:
![](https://cl.ly/3A3V2m272X0r/Image%202018-01-05%20at%2011.02.06%20AM.png)

takeover modal, note editor is disabled:
![](https://cl.ly/1e0B2w1P0t0s/Image%202018-01-05%20at%2011.04.57%20AM.png)

another user took over the post being edited modal, note editor is disabled:
![](https://cl.ly/2n1O3i3E321N/Edit_Post__WordPress_Local_Development...__WordPress_2018-01-05_11-10-53.jpg)

## Related Issues and/or PRs
https://github.com/WordPress/gutenberg/pull/4218

**Comment by adamsilverstein** [comment]
> One thing we should make sure is that the post locking works if someone is in the classic editor as well. 

This will work correctly as long as we use the same mechanism for locking for Gutenberg as for the classic editor. I tested this in my original PR (https://github.com/WordPress/gutenberg/pull/4217) that used the heartbeat API and included post locking by opening a post in Gutenberg, then trying to open in the classic editor and the other way around. 

I reopened my PR and will refresh and re-test to confirm.

**Comment by grappler** [comment]
I feel this feature should be added before 4.9.8 us released. Loosing changes could easily ruin people's experience with Gutenberg.

**Comment by adamsilverstein** [comment]
I'll update the PR.

**Comment by adamsilverstein** [comment]
@karmatosed from a design perspective, do you think an overlay and modal like we have in core currently is appropriate for the takeover interaction (see screenshots above). Or, do you have some other ideas howe we can implement these interactions: _when a user opens a post another user has locked_ and _when another user takes over editing of a post you are editing_. Do we have an existing method/instance where we lock Gutenberg and/or display an overlay? Media?

**Comment by karmatosed** [comment]
Great idea @adamsilverstein, here is my suggestion. I do wonder if it has to be a primary button to view posts or a link will do - happy to go with whatever works there, my leaning is towards a link.

![artboard 2](https://user-images.githubusercontent.com/253067/42708207-815c4600-86d4-11e8-9c27-081715e92b88.png)

**Comment by karmatosed** [comment]
Here is what the 3 buttons look like. I kind of wonder if there should be 3 buttons or links, I don't want to suggest a change in this that causes issues though. I did wonder about the order they are in, but again let's see maybe after about that, it's important to get it working for people.

![artboard 2](https://user-images.githubusercontent.com/253067/42708760-2557fdf2-86d6-11e8-8789-096886bd1b67.png)

**Comment by adamsilverstein** [comment]
Ok, I fixed the PR so the UI  more closely match the designs you presented here @karmatosed and posted updated screenshots: https://github.com/WordPress/gutenberg/pull/4217#issuecomment-414187220

Could still use a little CSS love to make it look great.

**Comment by davisshaver** [comment]
+1 to merging this before 4.9.8. Important if not critical for high volume publishing.

**Comment by eddhurst** [comment]
Wanted to add a +1 to this. Incredibly important to at least trigger the lock as soon as an editor opens the page to prevent unexpected loss of data.

**Comment by grappler** [comment]
With WordPress 5.0 RC expected in a month I am getting anxious that this is not going to be in Gutenberg in time for WordPress 5.0 release which would not be good.
